File: days/2/part1-solution.rb
################################################################################
#
#  Author:  Joe Cohen
#  Mood:    Frosty
#  Problem: https://adventofcode.com/2022/day/2
#  Status:  Solved
#
################################################################################
running_total = 0
file = File.open(__dir__ + '/input.txt', 'r')
# file = File.open(__dir__ + '/test-input-1.txt', 'r')
scores = { 'X' => 1, 'Y' => 2, 'Z' => 3 }

def calc_match_score(a, b)
  if (a == 'A' && b == 1) || (a == 'B' && b == 2) || (a == 'C' && b == 3)
    b + 3  #draw
  elsif (a == 'A' && b == 2) || (a == 'B' && b == 3) || (a == 'C' && b == 1)
    b + 6  #win
  else
    b  #loss
  end
end

while !file.eof?
  line = file.readline
  plays = line.split(' ')
  player_1 = plays[0]
  player_2 = scores[plays[1]]
  running_total += calc_match_score(player_1, player_2)
end

puts running_total

File: days/4/part1-solution.ts
/*******************************************************************************
*
*  Author:  Joe Cohen
*  Mood:    Chilly
*  Problem: https://adventofcode.com/2022/day/4
*  Status:  Solved
*
*******************************************************************************/
const nReadlines = require('n-readlines');
const inputData = new nReadlines(`${__dirname}/input.txt`);
// const inputData = new nReadlines(`${__dirname}/test-input.txt`);
let line, overlapsFound = 0;

while (line = inputData.next())
  overlapsFound += intervalsOverlap(line.toString()) ? 1 : 0;

console.log(`Overlaps found: ${overlapsFound}`);
const used = process.memoryUsage().heapUsed / 1024 / 1024;
console.log(`Memory used: ${Math.round(used * 100) / 100} MB`);

export {};

function intervalsOverlap(intervalString: string): boolean {
  const intervals = intervalString.split(',').map(interval => interval.split('-').map(Number));

  for (let i = 0; i < intervals.length; i++)
    for (let j = i + 1; j < intervals.length; j++)
      if (intervalsOverlapHelper(intervals[i], intervals[j]))
        return true;
  return false;
}

function intervalsOverlapHelper(interval1: number[], interval2: number[]): boolean {
  return (interval1[0] <= interval2[0] && interval2[1] <= interval1[1])
    || (interval2[0] <= interval1[0] && interval1[1] <= interval2[1]);
}